drivers/overlay/mount.go

```diff
@@ -77,12 +79,111 @@ func mountFromMain() {
 		fatal(err)
 	}
 
-	if err := os.Chdir(flag.Arg(0)); err != nil {
+	// Mount the arguments passed from the specified directory. Some of the
+	// paths mentioned in the values we pass to the kernel are relative to
+	// the specified directory.
+	homedir := flag.Arg(0)
+	if err := os.Chdir(homedir); err != nil {
 		fatal(err)
 	}
 
-	if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
-		fatal(err)
+	pageSize := unix.Getpagesize()
+	if len(options.Label) < pageSize {
+		if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
+			fatal(err)
+		}
+		os.Exit(0)
+	}
+
+	// Those arguments still took up too much space.  Open the diff
+	// directories and use their descriptor numbers as lowers, using
+	// /proc/self/fd as the current directory.
+
+	// Split out the various options, since we need to manipulate the
+	// paths, but we don't want to mess with other options.
+	var upper, work, lower, label, other string
+	for _, arg := range strings.Split(options.Label, ",") {
+		if strings.HasPrefix(arg, "upperdir=") {
```

Comment:
Style only:

```go
nv := strings.SplitN(arg, "=")
switch nv[0] {
case "upperdir":
    upper = nv[1]
…
}
```
(and eventually `strings.Cut` with Go 1.18.)

---

Separately: The caller has all of this data. It feels a bit tedious to format it into a single string, and then parse it out again — but changing that would be a rather larger rewrite.